adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/lwm2m/LwM2MResourceDirectory.java

```diff
@@ -0,0 +1,694 @@
+                .withTag(TracingHelper.TAG_DEVICE_ID, ctx.getOriginDevice().getDeviceId())
+                .withTag(TAG_LWM2M_REGISTRATION_ID, registrationId)
+                .start();
+
+        LOG.debug("processing LwM2M registration update request [tenant-id: {}, device-id: {}, registration ID]",
+                ctx.getTenantId(), ctx.getOriginDevice().getDeviceId(), registrationId);
+        final Map<String, Object> items = new HashMap<>(2);
+        Optional.ofNullable(bindingMode).ifPresent(bm -> items.put("binding mode", bm.name()));
+        Optional.ofNullable(lifetime).ifPresent(lt -> items.put("lifetime (secs)", lt));
+        currentSpan.log(items);
+
+        if (lifetime != null && lifetime <= 0) {
+            result.fail(new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST,
+                    "lifetime query parameter must be positive integer"));
+        } else if (bindingMode != null && !supportedBindingModes.contains(bindingMode)) {
+            result.fail(new ClientErrorException(HttpURLConnection.HTTP_PRECON_FAILED,
+                    "server supports binding modes U and UQ only"));
+        } else {
+            final var registrationUpdate = new RegistrationUpdate(
+                    registrationId,
+                    EndpointContextUtil.extractIdentity(ctx.getExchange().advanced().getRequest().getSourceContext()),
+                    lifetime,
+                    null,
+                    bindingMode,
+                    objectlinks,
+                    null);
+            final var updatedRegistration = lwm2mRegistrationStore.updateRegistration(registrationUpdate);
+
+            sendLifetimeAsTtdEvent(ctx, updatedRegistration.getUpdatedRegistration(), currentSpan)
+                    .onSuccess(ok -> result.complete(ResponseCode.CHANGED))
+                    .onFailure(t -> result.fail(t));
+        }
+        return result.future()
+                .onFailure(t -> TracingHelper.logError(currentSpan, t))
+                .onComplete(r -> currentSpan.finish());
+    }
+
+    private Future<?> sendLifetimeAsTtdEvent(
+            final CoapContext ctx,
+            final Registration registration,
+            final Span currentSpan) {
+
+        final int lifetime = registration.getBindingMode() == BindingMode.U
+                ? registration.getLifeTimeInSec().intValue()
+                : 20;
+        return getAdapter().sendTtdEvent(
+                ctx.getTenantId(),
+                ctx.getOriginDevice().getDeviceId(),
+                ctx.getAuthenticatedDevice(),
+                lifetime,
+                ctx.getTracingContext());
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Handles a request to delete a device's registration.
+     */
+    @Override
+    protected Future<?> handleDelete(final CoapContext ctx) {
+
+        final List<String> uri = ctx.getExchange().getRequestOptions().getUriPath();
+
+        if (uri.size() >= 2) {
```

Comment:
`handlePost` is more strict about the `uri.size()`. Is that intended?

Reply:
I will change it to only accept URIs with two path segments ....